File: sdks/api/src/main/java/com/vivid/clients/api/metadata/JsonMetadataValue.java
package com.vivid.clients.api.metadata;

import com.vivid.clients.api.MetadataValue;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;

public record JsonMetadataValue(JsonNode content) implements MetadataValue<JsonNode> {
    @Override
    public JsonNode getContent() {
        return content;
    }

    public <T> T convertTo(ObjectMapper objectMapper, Class<T> type) {
        return objectMapper.convertValue(content, type);
    }

    public <T> T readAs(ObjectMapper objectMapper, JsonMetadataValue metadataValue, Class<T> type) {
        return metadataValue.convertTo(objectMapper, type);
    }
}

File: sdks/kotlin/src/main/kotlin/com/vivid/sdk/api/metadata/JsonMetadataValue.kt
package com.vivid.sdk.api.metadata

import tools.jackson.databind.JsonNode

import com.vivid.sdk.api.MetadataValue
import tools.jackson.databind.ObjectMapper
import kotlin.reflect.KClass

/**
 * Metadata value containing a string that is a JSON.
 *
 * @property content the JSON string
 */
data class JsonMetadataValue(
    val content: JsonNode
) : MetadataValue {
    fun <T: Any>convertTo(objectMapper: ObjectMapper, clazz: KClass<T>): T {
        return objectMapper.treeToValue(content, clazz.java)
    }
}

inline fun <reified T: Any>JsonMetadataValue.convertTo(objectMapper: ObjectMapper): T {
    return convertTo(objectMapper, T::class)
}

fun <T: Any>ObjectMapper.readAs(metadata: JsonMetadataValue, type: KClass<T>): T {
    return metadata.convertTo(this, type)
}

inline fun <reified T: Any>ObjectMapper.readAs(metadata: JsonMetadataValue): T {
    return readAs(metadata, T::class)
}
